Fix 2128

pygy opened the pull request:
This fixes #2128, a DOM mocks bug that I discovered while fixing #2128, and removes quite a bit of code.

Change log item coming as soon as I have the PR number.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated `docs/change-log.md`

pygy commented on render/render.js:
```diff
@@ -312,22 +315,22 @@ module.exports = function($window) {
-							updateNode(parent, o, v, hooks, getNextSibling(old, oldEnd + 1, nextSibling), ns)
-							insertNode(parent, toFragment(v), nextSibling)
+							insertNode(parent, toFragment(o), nextSibling)
+							updateNode(parent, o, v, hooks, nextSibling, ns)
```
Just below, replacing `o.skip = true` with `old[oldIndex] = null` works as well, and it would save us a vnode slot.

Is there a reason not to mutate the old children list?

tivac replied:
> save us a vnode slot

Not sure I follow...

pygy replied:
Sorry, that wasn't clear. If we do that, we can remove the `.skip` field on vnodes, improving their memory footprint.

tivac commented on render/render.js:
```diff
@@ -51,18 +51,17 @@ module.exports = function($window) {
 			vnode.state = {}
 			if (vnode.attrs != null) initLifecycle(vnode.attrs, vnode, hooks)
 			switch (tag) {
-				case "#": return createText(parent, vnode, nextSibling)
```
I don't have a strong opinion one way or another, but why'd this change?

pygy replied:
The individual `createX` methods used to return a DOM node that was used in `createComponent()` and `updateNodes()` (in the map-based diff section).

This isn't needed anymore (and wasn't necessary in the first place AFAICT,  since `insertNode()` in `createComponent()` was redundant and we can get `v.dom` in `updateNodes()`).

So I made it explicit that `createNode()` wasn't returning anything.

tivac commented:
This seems... reasonable. This is definitely the parts of mithril I understand the least though so I'm hesitant to actually post a review.

pygy commented:
The main change here is the simplification to get the nextSibling.

When processing both lists downwards (unkeyed diff, first part of the keyed diff), the bottom part of the list hasn't been updated yet, and we need to look for the nextSibling in the  `old` list.

When processing the new list upwards (the other three sections of the keyed list algo), the DOM nodes below `end` have already been updated, and we can set `nextSibling` to `v.dom` if it isn't null.

I've changed the list reversal part of the diff so that it processes the old list top-down and the new list bottom up to take advantage of that.

I've also moved the `insertNode()` calls before `updateNode()` where the nodes actually move, to take advantage of the cached `nextSibling`.

All in all, It saves us many `getNextSibling()` calls. I should probably update the comment about this to make it more clear.

pygy commented:
@tivac I just pushed another commit that fully disentangles keyed and unkeyed diff.

With that, the pool stuff removed and the bug fixes, `updateNodes()` is at its simplest ever (the parts that I had a hard time to understand were actually buggy). It may be a good time to re-read it top-down.

tivac commented:
I'll take a look tomorrow

pygy commented:
Ok, I've cleaned it up some more, and added some comments :-)

pygy commented:
@isiahmeadows do you want to be removed from the owners file as well (that was an automatic review request)?

isiahmeadows commented:
@pygy [I already took care of that](https://github.com/MithrilJS/mithril.js/commit/0dbf17c6212ed1ed3830df46ae414182d5ddfdc0).

pygy commented:
:+1: I had missed it.

tivac commented on render/render.js:
```diff
@@ -256,84 +265,107 @@ module.exports = function($window) {
-			for(; start < commonLength; start++) {
-				if (old[start] != null && vnodes[start] != null) {
-					if (old[start].key == null && vnodes[start].key == null) isUnkeyed = true
+			// default to keyed because, when either list isfull of null nodes, it has fewer branches
```
>isfull

"is full"

pygy commented:
Here we go!

pygy commented:
I think most of the updateNodes fixes could be backported, I'll have to review the patches (I'm quite busy at the moment though so I can't promise it'll be quick).

pygy commented:
@JacksonJN actually there are some changes in the `next` branch that are backwards incompatible (state/hooks-related), and disentangling them is more trouble than I can deal with now, so this will go in v2.

JacksonJN commented:
@pygy Is v2 currently planned or is there an ETA?

pygy commented:
We're actively working on it (there was a hiatus).

JacksonJN commented:
Thanks for the info.
